keps/sig-testing/0014-20180523-test-cluster-framework.md

```diff
@@ -0,0 +1,169 @@
+
+There is a clear tension between Alice and Carol’s frustrations. Left to their
+own devices, Alice is likely to contribute to the proliferation of test
+infrastructure that frustrates Carol, and Carol may find herself advocating
+that Alice be prevented from acting to resolve her own frustration with the
+status quo.
+
+Most of the content of this KEP has already [been worked on and fleshed
+out][the_doc] mostly by people from sig-testing, more specifically people
+working on the [testing-commons subproject][testing_commons_notes].
+
+[key_personas]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.czu8o1uor9qk
+[alice]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.17sl3t5a8nu
+[bob]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.mecveaw12b4d
+[carol]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.4i5imqvwxv3a
+[testing_commons_notes]: https://docs.google.com/document/d/1TOC8vnmlkWw6HRNHoe5xSv5-qv7LelX6XK3UVCHuwb0/edit#
+[the_doc]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#
+
+## Motivation
+
+With this proposal we would like to address these three points of friction:
+- For Carol, we introduce an abstraction which is intended to unify existing
+  and future methods of spinning up test clusters.
+- For Alice’s frustration, we build a lightweight testing framework with no
+  dependency on `k8s.io/kubernetes`
+- For the tension between Alice and Carol, we ensure that Alice’s framework is
+  fully accessible through Carol’s abstraction.
+
+
+<!--
+TODO: Add note that we did interviews and talked to different people, noted in
+      the "Validation" section of the personas?
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+1. To design an abstraction of “standing up a kubernetes cluster” which is:
+    - expressive enough that most or all e2e tests ought to be able to use it,
+      rather than standing things up directly
+    - abstract enough that it could in principle be wrapped around many existing
+      methods of standing up clusters
+    - equally easy to use interactively and in CI
+2. To build a lightweight testing framework for CLI use which
+    - is not dependant on `k8s.io/kubernetes`
+    - is easy for newcomers to use
+    - makes use of the abstraction above
+3. To build a heavier weight testing framework, for API/Controller use which
+    - is fully functional
+    - makes use of the abstraction above
+
+
+### Non-Goals
+
+- To wrap any existing “ways of standing up a cluster” in our abstraction
+   - Of course, wrapping existing technology may well be the most efficient way
+     to meet goals 2 and 3 (building CLI and API/Controller testing frameworks). 
+- To port large numbers of existing tests to use our abstraction.
```

Comment:
absent this, are we just creating a new standard?

Reply:
To understand if this framework is useful we were thinking of porting a couple of tests for each (or most) of the test suites. An example of how such a port could look can be found [here](https://github.com/pivotal-k8s/kubernetes/compare/master...pivotal-k8s:translating-test-cmd-to-cluster-api#diff-efd84624607a7f9efbb2aaeeb004557aR10), where we experimented with porting different tests from `test-cmd.sh`.  And I believe we should port not only tests from `test-cmd.sh` but also other.

But porting *all* the tests is not in the scope of this KEP.